File: src/simulation/culture/aspect/dependency/LineDependency.java
package simulation.culture.aspect.dependency;

import kotlin.Pair;
import simulation.culture.aspect.*;
import simulation.culture.group.AspectCenter;
import simulation.culture.group.Group;
import simulation.culture.group.request.ResourceEvaluator;
import simulation.space.resource.MutableResourcePack;
import simulation.space.resource.tag.ResourceTag;

import java.util.Objects;
import java.util.stream.Collectors;

public class LineDependency extends AbstractDependency {
    private boolean isAlreadyUsed = false;
    private Pair<ConverseWrapper, ConverseWrapper> line;

    public LineDependency(ResourceTag tag, Pair<ConverseWrapper, ConverseWrapper> line) {
        super(tag);
        this.line = line;
    }

    @Override
    public String getName() {
        return line.getFirst().getName() + " from " + line.getSecond().getName();
    }

    public Pair<ConverseWrapper, ConverseWrapper> getLine() {
        return line;
    }

    @Override
    public boolean isCycleDependency(Aspect aspect) {
        if (isAlreadyUsed) {
            return false;
        }
        isAlreadyUsed = true;
        boolean b = line.getSecond().getDependencies().values().stream().anyMatch(dependencies -> dependencies.stream()
                .anyMatch(dependency -> dependency.isCycleDependencyInner(aspect))) || (line.getSecond().equals(aspect));
        isAlreadyUsed = false;
        return b;
    }

    @Override
    public boolean isCycleDependencyInner(Aspect aspect) {
        if (isAlreadyUsed) {
            return false;
        }
        isAlreadyUsed = true;
        boolean b = line.getSecond().getDependencies().values().stream().anyMatch(dependencies -> dependencies.stream()
                .anyMatch(dependency -> dependency.isCycleDependencyInner(aspect))) || (line.getSecond().equals(aspect));
        isAlreadyUsed = false;
        return b;
    }

    @Override
    public AspectResult useDependency(AspectController controller) {
        try {
            MutableResourcePack resourcePack = new MutableResourcePack();
            if (isAlreadyUsed || controller.getCeiling() <= 0 || !foodForInsertMeaning() && controller.isMeaningNeeded()) {
                return new AspectResult(resourcePack, null);
            }
            isAlreadyUsed = true;
            AspectResult _p = line.getSecond().use(new AspectController(
                            controller.getCeiling(),
                            controller.getFloor(),
                            new ResourceEvaluator(
                                    rp -> rp.getPackedResource(line.getFirst().resource),
                                    rp -> rp.getAmount(line.getFirst().resource)
                            ),
                            controller.getPopulationCenter(),
                            controller.getTerritory(),
                            controller.isMeaningNeeded(),
                            controller.getMeaning()
                    ));
            resourcePack.addAll(_p.resources.getPackedResource(line.getFirst().resource).getResources().stream()
                    .flatMap(res -> res.applyAndConsumeAspect(line.getFirst().aspect, controller.getCeiling()).stream())
                    .collect(Collectors.toList()));
            resourcePack.addAll(_p.resources);
            isAlreadyUsed = false;
            return new AspectResult(_p.isFinished, resourcePack, null);
        } catch (NullPointerException e) {
            throw new RuntimeException("No such aspect in Group");
        }
    }

    private boolean foodForInsertMeaning() {
        return !tag.equals(ResourceTag.phony()) || getNextWrapper().canInsertMeaning;
    }

    public ConverseWrapper getNextWrapper() {
        return line.getSecond();
    }

    @Override
    public LineDependency copy() {
        return new LineDependency(tag, line);
    }

    @Override
    public void swapDependencies(AspectCenter aspectCenter) {
        line = new Pair<>(
                (ConverseWrapper) aspectCenter.getAspectPool().get(line.getSecond()),
                (ConverseWrapper) aspectCenter.getAspectPool().get(line.getSecond())
        );
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        LineDependency that = (LineDependency) o;
        return Objects.equals(line, that.line);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), line);
    }
}

File: src/simulation/culture/aspect/dependency/LineDependency.kt
package simulation.culture.aspect.dependency

import simulation.culture.aspect.Aspect
import simulation.culture.aspect.AspectController
import simulation.culture.aspect.AspectResult
import simulation.culture.aspect.ConverseWrapper
import simulation.culture.group.centers.AspectCenter
import simulation.culture.group.request.resourceEvaluator
import simulation.space.resource.container.MutableResourcePack
import java.util.*

class LineDependency(
        isPhony: Boolean,
        private var parentConverseWrapper: ConverseWrapper,
        var converseWrapper: ConverseWrapper
) : AbstractDependency(isPhony) {
    private var isAlreadyUsed = false
    override val name: String
        get() = "from ${converseWrapper.name}"

    override fun isCycleDependency(otherAspect: Aspect): Boolean {
        if (parentConverseWrapper == converseWrapper) return true
        if (parentConverseWrapper.used) return false
        parentConverseWrapper.used = true
        val b = converseWrapper == otherAspect
        parentConverseWrapper.used = false
        return b
    }

    override fun isCycleDependencyInner(otherAspect: Aspect) = isCycleDependency(otherAspect)

    override fun useDependency(controller: AspectController): AspectResult {
        return try {
            val resourcePack = MutableResourcePack()
            if (isAlreadyUsed || controller.ceiling <= 0 || !goodForInsertMeaning() && controller.isMeaningNeeded) {
                return AspectResult(resourcePack, null)
            }
            isAlreadyUsed = true
            val _p = converseWrapper.use(controller.copy(
                    evaluator = resourceEvaluator(parentConverseWrapper.resource)
            ))
            resourcePack.addAll(_p.resources.getResource(parentConverseWrapper.resource).resources.flatMap {
                        it.applyActionAndConsume(
                                parentConverseWrapper.aspect.core.resourceAction,
                                controller.ceiling,
                                true
                        )
                    })
            resourcePack.addAll(_p.resources)
            isAlreadyUsed = false
            AspectResult(_p.isFinished, emptyList(), resourcePack, null)
        } catch (e: NullPointerException) {
            isAlreadyUsed = false
            throw RuntimeException("No such aspect in Group")
        }
    }

    private fun goodForInsertMeaning() = !isPhony || converseWrapper.canInsertMeaning

    override fun copy() = LineDependency(isPhony, parentConverseWrapper, converseWrapper)

    override fun swapDependencies(aspectCenter: AspectCenter) {
        parentConverseWrapper = aspectCenter.aspectPool.getValue(parentConverseWrapper) as ConverseWrapper
        converseWrapper = aspectCenter.aspectPool.getValue(converseWrapper) as ConverseWrapper
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        if (!super.equals(other)) return false
        val that = other as LineDependency
        return parentConverseWrapper == that.parentConverseWrapper
                && converseWrapper == that.converseWrapper
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), parentConverseWrapper, converseWrapper)
    }

}